### PR TITLE
Add support for related fields in generic `IndexView.list_display`

### LINKED ISSUE #11494: Using a related field in Snippet/ModelViewSet causes exception when index list requested
### Issue Summary
When referencing a related model field in a SnippetViewSet or ModelViewSet, requesting the index view causes an exception because the column label cannot be determined.

Specifically, it fails within the Django `label_for_field()` function:

```
Traceback (most recent call last):
  File ".../.venv/lib/python3.10/site-packages/django/db/models/options.py", line 673, in get_field
    return self.fields_map[field_name]
KeyError: 'full_featured.text'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../.venv/lib/python3.10/site-packages/django/contrib/admin/utils.py", line 333, in label_for_field
    field = _get_non_gfk_field(model._meta, name)
  File ".../.venv/lib/python3.10/site-packages/django/contrib/admin/utils.py", line 302, in _get_non_gfk_field
    field = opts.get_field(name)
  File ".../.venv/lib/python3.10/site-packages/django/db/models/options.py", line 675, in get_field
    raise FieldDoesNotExist(
django.core.exceptions.FieldDoesNotExist: SnippetChooserModel has no field named 'full_featured.text'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../wagtail/snippets/tests/test_viewset.py", line 748, in test_snippetchoosermodel_related_fields
    response = self.client.get(self.get_url("list"))
  File ".../.venv/lib/python3.10/site-packages/django/test/client.py", line 836, in get
    response = super().get(path, data=data, secure=secure, **extra)
  File ".../.venv/lib/python3.10/site-packages/django/test/client.py", line 424, in get
    return self.generic(
  File ".../.venv/lib/python3.10/site-packages/django/test/client.py", line 541, in generic
    return self.request(**r)
  File ".../.venv/lib/python3.10/site-packages/django/test/client.py", line 810, in request
    self.check_exception(response)
  File ".../.venv/lib/python3.10/site-packages/django/test/client.py", line 663, in check_exception
    raise exc_value
  File ".../.venv/lib/python3.10/site-packages/django/core/handlers/exception.py", line 56, in inner
    response = get_response(request)
  File ".../.venv/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File ".../.venv/lib/python3.10/site-packages/django/views/decorators/cache.py", line 62, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File ".../wagtail/admin/urls/__init__.py", line 180, in wrapper
    return view_func(request, *args, **kwargs)
  File ".../wagtail/admin/auth.py", line 167, in decorated_view
    response = view_func(request, *args, **kwargs)
  File ".../.venv/lib/python3.10/site-packages/django/views/generic/base.py", line 78, in view
    self.setup(request, *args, **kwargs)
  File ".../wagtail/admin/views/generic/models.py", line 111, in setup
    self.columns = self.get_columns()
  File ".../wagtail/snippets/views/snippets.py", line 180, in get_columns
    *super().get_columns(),
  File ".../wagtail/admin/views/generic/models.py", line 336, in get_columns
    column = self._get_custom_column(field)
  File ".../wagtail/admin/views/generic/models.py", line 309, in _get_custom_column
    label, attr = label_for_field(field_name, self.model, return_attr=True)
  File ".../.venv/lib/python3.10/site-packages/django/contrib/admin/utils.py", line 361, in label_for_field
    raise AttributeError(message)
AttributeError: Unable to lookup 'full_featured.text' on SnippetChooserModel
```

### Steps to Reproduce

1. Clone the wagtail branch at https://github.com/dkirkham/wagtail/commit/e942cae55fe7e8324eca942b29638c2af9551f48 within a working Wagtail development environment
2. Run the test: `python runtests.py wagtail.snippets.tests.test_viewset.TestRelatedFieldColumns`
3. Note the crash backtrace in the test report.

### Background
I've found myself wanting to refer to fields within related models within ModelViewSet index lists. This is because the `__str__` methods in those referenced models often include other contextual information which is useful in other circumstances, but too much for the index list (which often has that context in adjacent columns). I'd also like to avoid creating extra methods in models to access related model fields.

I (perhaps naïvely) tried using `list_display = ["field__related_field"]` a format which works in `search_fields` and `ordering`, but this fails. Digging into the code, `list_display` ultimately uses `multigetattr()` as the data accessor which reproduces template semantics using dot-separated fields: `list_display = ["field.related_field"]`. Neither format however works to automatically obtain a column label.

It is possible to use the `Column()` class to directly provide a label: `list_display = [Column("field.related_field", label="Related Field")]`

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: (no)
-   I have confirmed that this issue can be demonstrated in a Wagtail test case: (yes)

Is it a bug? The fact that there is code to allow access to the related field data, but the code fails trying to work out the column name would suggest so. The generality of the `multigetattr()` data accessor might be difficult to reproduce when finding the label, so the workaround using `Column()` may be an acceptable and more dependable solution. If so, it might be useful to both detect the failure to find a label and raise this as a warning or configuration error, rather than failing when the index list is requested; and to document the preferred solution.

### Technical details

-   Python version: 3.10, but likely not relevant.
-   Django version: 4.0.10 and 4.2.8, probably others.
-   Wagtail version: 5.2 and wagtail#main
-   Browser version: irrelevant

### Working on this

I've started a [branch](https://github.com/dkirkham/wagtail/tree/fix-index-view-get-related-field-label) with the test demonstrating the failure. I haven't yet started looking to see if an automatic method of determining the label for a related field is possible.

Anyone can contribute to this. View our [contributing guidelines](https://docs.wagtail.org/en/latest/contributing/index.html), add a comment to the issue once you’re ready to start.

### PR DESCRIPTION
**Pull request by abdlrhman08**
Fixes  #11494 

After:
image

Also works for recursive relations

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Comment by lb-** [comment]
Thanks for this PR @abdlrhman08  I have started the CI and requested a review from @laymonage

File: wagtail/admin/views/generic/models.py
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:274]: I don't think this is necessary. Or if we want to, we definitely don't want to use the field's internal name directly here, which is a code string in snake_case and should not be exposed to the user. A better idea would be to use the model's verbose name (`model._meta.verbose_name`)

File: wagtail/admin/views/generic/models.py
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:265]: Is there a reason why we use `getattr()` here instead of `model_class._meta` when we're always accessing `_meta` and we're not using a default fallback value?

File: wagtail/admin/views/generic/models.py
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:259]: Can we split by `django.db.models.constants.LOOKUP_SEP` instead?

File: wagtail/admin/views/generic/models.py
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:281]: Does this work? I imagine you'd need double underscores for this to work. If we used `LOOKUP_SEP` instead of dot-separated fields, we probably wouldn't need this.

**Review by laymonage — CHANGES_REQUESTED** [review]
Thanks for giving this a go! I added some thoughts on this feature in https://github.com/wagtail/wagtail/issues/11494#issuecomment-1948769490. Could you have a look (and perhaps also see how Django's ModelAdmin implements it and how it works over there)?

File: wagtail/admin/views/generic/models.py
Comment (abdlrhman08) [reply wagtail/admin/views/generic/models.py]: Changed

File: wagtail/admin/views/generic/models.py
Comment (abdlrhman08) [reply wagtail/admin/views/generic/models.py]: Tbh, I don't remember why I used `getattr()` instead of directly accessing it using `model_class._meta`. I changed it to `model_class._meta`.

File: wagtail/admin/views/generic/models.py
Comment (abdlrhman08) [reply wagtail/admin/views/generic/models.py]: Yes you are totally right, if `LOOKUP_SEP` is used there is no need for this.

File: wagtail/admin/views/generic/models.py
Comment (abdlrhman08) [reply wagtail/admin/views/generic/models.py]: Changed to model's verbose name, can be removed if not necessary.

**Comment by abdlrhman08** [comment]
Thank you for your review @laymonage ! I took a look at how Django's ModelAdmin implemented it and, I have changed the code according to requested changes and the tests accordingly. Can you have another look on the changes ? 

There is still one thing I would like you to take a look at, in this [line here](https://github.com/abdlrhman08/wagtail/blob/3758a71b2b425bb17910279b00f905602756a4d5/wagtail/admin/views/generic/models.py#L285), I had to build the dotted path again, because `multigetattr()` uses it in the column to get the data in the field. Should we change `multigetattr` to also split by double underscores or build the dotted field path again as I did.

I am happy to change if any changes are requested !

File: wagtail/admin/views/generic/models.py
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:272]: This can be very long if we traverse several relations, maybe just take the `verbose_name` of the last relation?

```suggestion
            foreign_field_name = foreign_field.verbose_name + " "
```

File: wagtail/snippets/tests/test_viewset.py
Comment (laymonage) [inline wagtail/snippets/tests/test_viewset.py:785]: Could you also assert that the response contains the text of the full featured snippet, as well as the header correctly showing the label (foreign key's verbose_name + the field's verbose_name)

File: wagtail/snippets/tests/test_viewset.py
Comment (laymonage) [inline wagtail/snippets/tests/test_viewset.py:775]: Could you add another test case where the foreign key is null? e.g.

```py
    def test_snippetchoosermodel_related_fields(self):
        advert = Advert(url="https://example.com/free_examples", text="Free Examples")
        advert.save()
        scm = SnippetChooserModel(advert=advert)
        scm.save()

        response = self.client.get(self.get_url("list"))

        self.assertEqual(response.status_code, 200)
        # assert the header label here
        # (and maybe assert that the cell is empty)
```

Right now it seems to crash as the `multigetattr` function tries to access `None["text"]`. You may have to add a `try`/`except` in `Column.get_value()` to return `None` when this occurs.

**Review by laymonage — COMMENTED** [review]
Thanks, this is looking great and I think it's almost there!

File: wagtail/admin/views/generic/models.py
Comment (abdlrhman08) [reply wagtail/admin/views/generic/models.py:272]: Yeah, I think taking the last relations name only is enough. I wrote like this in the beginning, because in Django Admin it displays the complete relation

File: wagtail/snippets/tests/test_viewset.py
Comment (abdlrhman08) [reply wagtail/snippets/tests/test_viewset.py:775]: Done

File: wagtail/snippets/tests/test_viewset.py
Comment (abdlrhman08) [reply wagtail/snippets/tests/test_viewset.py:785]: Done

**Comment by abdlrhman08** [comment]
Thank you for your review and help @laymonage ! I added one more test case. At the moment, I can't think of more test cases. I am happy to add any corner cases or more tests.

File: wagtail/test/testapp/wagtail_hooks.py
Comment (laymonage) [inline wagtail/test/testapp/wagtail_hooks.py:389]: Sorry I forgot to suggest this before, could we add another case for multi-level relations? e.g.

```suggestion
    list_display = ["__str__", "full_featured__text", "full_featured__latest_revision__created_at"]
```

In the test, you'll need to do `ffs.save_revision()` so the `latest_revision` is populated. The method returns the `Revision` object, so you can hold on to it to get the `created_at` for assertion. You can then use `wagtail.utils.timestamps.render_timestamp` to assert the rendered timestamp is in the response. The label should be "Latest revision created at".

**Review by laymonage — COMMENTED** [review]
Thank you, this is looking great! One small comment to improve the label for other languages, and another comment to ensure that this feature works with multi-level relations. I think we'll be good to go after that!

**Comment by abdlrhman08** [comment]
I just went through the code of `label_for_field`, and I found out that if we are using `LOOKUP_SEP` in `list_diplay` there is no need for my addition, since `label_for_field` will just work out of the box. I'm sorry for noticing this late. So basically this is enough to do it:
```python
def _get_custom_column(self, field_name, column_class=Column, **kwargs):
        label, attr = label_for_field(field_name, self.model, return_attr=True)
        sort_key = getattr(attr, "admin_order_field", None)

        # attr is None if the field is an actual database field,
        # so it's possible to sort by it
        if attr is None:
            sort_key = field_name

        if not hasattr(self.model, field_name):
            sort_key = field_name
            field_name = field_name.replace(LOOKUP_SEP, ".")
            
        return column_class(
            field_name,
            label=capfirst(label),
            sort_key=sort_key,
            **kwargs,
        )
```

The only problem here is no translation and we will have very long relation name.
let me know what you think about this

File: wagtail/admin/views/generic/models.py
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:290]: If we didn't immediately unpack the splitting result into `*relation, field`, we can reuse it here so we don't have to construct a new list. It's a very small optimisation but I think it's nice to have

File: wagtail/snippets/tests/test_viewset.py
Comment (laymonage) [inline wagtail/snippets/tests/test_viewset.py:789]: If we add a `verbose_name` to the `ForeignKey`, we can make a more specific assertion here to ensure that we're using the verbose name of the field rather than just the field name

File: wagtail/snippets/tests/test_viewset.py
Comment (laymonage) [inline wagtail/snippets/tests/test_viewset.py:784]: It's best to move the creation of this object to the test for empty FK specifically, as otherwise the other unrelated tests could also fail if the implementation didn't handle the empty FK case.

File: wagtail/snippets/tests/test_viewset.py
Comment (laymonage) [inline wagtail/snippets/tests/test_viewset.py:786]: We're not testing the `SnippetChooserModel`, so it's better to use a more descriptive name e.g.

```suggestion
    def test_single_level_relation(self):
```

**Review by laymonage — COMMENTED** [review]
Thanks @abdlrhman08, this is looking great! I only have a few minor comments. I'll make the changes as I merge.

> I just went through the code of label_for_field, and I found out that if we are using LOOKUP_SEP in list_diplay there is no need for my addition, since label_for_field would just work out of the box. I'm sorry for noticing this late. So basically this is enough to do it:

No worries, you're right but that only applies to the current in-development version of Django because https://github.com/django/django/pull/17357 is quite recent. If you install the latest release, you'll get an `AttributeError` when trying to do so, e.g.

```
AttributeError: Unable to lookup 'favorite_book__author__name' on Person
```

We currently support Django 4.2 and 5.0, so we still need to provide our own implementation anyway. This also lets us to do some tweaks like only using the final related model, and also adding the ability for other languages to swap the order of the model name vs the field name.

**Review by laymonage — APPROVED** [review]

**Comment by laymonage** [comment]
Thanks! I've added your name to the contributors list as `Abdelrahman Hamada`. Let me know if you'd like a different name listed.

**Comment by abdlrhman08** [comment]
Thank you ! looking forward to helping the community more

**Pull request merged**
